src/main/scala/dsptools/numbers/UIntTypeclasses.scala

```diff
@@ -3,99 +3,116 @@
+    // TODO: Saturating mux should be outside of ShiftRegister
+    val sum = {
+      if(context.overflowType == Grow) f +& g
+      else if (context.overflowType == Wrap)  f +% g
+      else throw DspException("Saturating add hasn't been implemented")
     }
+    ShiftRegister(sum, context.numAddPipes)
   }
-  def times(f: UInt, g: UInt): UInt = {
-    f * g
-  }
-  def one: UInt = BigInt(1).U
-  def zero: UInt = BigInt(0).U
-  def negate(f: UInt): UInt = -f
   override def minus(f: UInt, g: UInt): UInt = {
-    if(context.overflowType == Grow) {
-      f -& g
-    } else if (context.overflowType == Saturate) {
-      println("Saturating subtract is broken right now!")
-      val grow = f -& g
-      val nogrow = f -% g
-      val max = 3.U
-      val min = 0.U
-
-      Mux(grow === nogrow,
-        nogrow,
-        Mux(grow > 0.U, max, min)
-      )
-    } else {
-      f -% g
+    val diff = {
+      if(context.overflowType == Grow) f -& g
+      else if (context.overflowType == Wrap) f -% g
+      else throw DspException("Saturating subtractor hasn't been implemented")
     }
+    ShiftRegister(diff, context.numAddPipes)
   }
-}
-
-trait UIntImpl {
-  implicit object UIntIntegerImpl extends UIntInteger
-  implicit def fromInt(x: Double): UInt = x.toInt.U
+  def negate(f: UInt): UInt = throw DspException("Can't negate UInt and get UInt")
+  def times(f: UInt, g: UInt): UInt = {
+    ShiftRegister(f * g, context.numMulPipes)
+  }  
 }
 
 trait UIntOrder extends Any with Order[UInt] with hasContext {
   override def compare(x: UInt, y: UInt): ComparisonBundle = {
     ComparisonHelper(x === y, x < y)
   }
+  override def eqv(x: UInt, y: UInt): Bool = x === y
+  override def neqv(x: UInt, y:UInt): Bool = x =/= y
+  override def lt(x: UInt, y: UInt): Bool = x < y
+  override def lteqv(x: UInt, y: UInt): Bool = x <= y
+  override def gt(x: UInt, y: UInt): Bool = x > y
+  override def gteqv(x: UInt, y: UInt): Bool = x >= y
+  // min, max depends on lt, gt
 }
 
 trait UIntSigned extends Any with Signed[UInt] with hasContext {
   def signum(a: UInt): ComparisonBundle = {
     ComparisonHelper(a === 0.U, a < 0.U)
   }
-
-  /** An idempotent function that ensures an object has a non-negative sign. */
   def abs(a: UInt): UInt = a // UInts are unsigned!
+  override def isSignZero(a: UInt): Bool = a === 0.U
+  override def isSignPositive(a: UInt): Bool = !isSignZero(a)
+  override def isSignNegative(a: UInt): Bool = false.B
+  // isSignNonZero, isSignNonPositive, isSignNonNegative derived from above (!)
 }
+
 trait UIntIsReal extends Any with IsIntegral[UInt] with UIntOrder with UIntSigned with hasContext {
   def toDouble(a: UInt): DspReal = ???
+  // In IsIntegral: ceil, floor, round, truncate (from IsReal) already defined as itself; 
+  // isWhole always true
+  
+  def isOdd(a: UInt): Bool = a(0)
+  // isEven derived from isOdd
+  // Note: whatever Chisel does -- is there overflow?
+  // Generally better to use your own mod if you know input bounds
```

Comment:
Should this be done in chisel or firrtl? It would be cool if you used the range API and a firrtl pass replaced a mod with a Mux(test, a, a - b) or whatever